docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+# Artifact Outputs for Incremental Generators
+
+## Summary
+
+This proposal adds a mechanism that allows [incremental generators](incremental-generators.md) to output non-source artifacts, that is, text or binary files that are emitted to disk alongside the compiled assembly.
+
+### Rationale
+
+There exists a desire for incremental source generators be able to write out a set of accompanying files, alongside the generated sources that they add to the compilation. An example of this is JavaScript interop with .NET code. When exposing a .NET method to JavaScript today it is possible to generate the C# interop source code that enables the call to be made from JavaScript, but the user must still hand author the corresponding JavaScript code. Instead, with this feature the incremental generator could emit both the required C# code alongside a JavaScript file that provides the equivalent JavaScript interop code.
+
+### High Level Design Goals
+
+- Allow Incremental Generators to emit source or binary files to disk alongside the emitted .NET assembly
+- The user can control where the generated artifacts are emitted with a suitable default
+- Be optimal in terms of file writes / copies and ensure we keep the correct incremental build semantics for MSBuild
+- Do *not* run as part of the IDE experience. Files are only emitted to disk as part of command line compilation
+
+## Implementation
+
+The feature is exposed to incremental generator authors via a new output API:
+
+```csharp
+namespace Microsoft.CodeAnalysis;
+
+public readonly partial struct IncrementalGeneratorInitializationContext
+{
+    public void RegisterArtifactOutput<TSource>(IncrementalValueProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+    public void RegisterArtifactOutput<TSource>(IncrementalValuesProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+}
+
+public readonly struct ArtifactProductionContext
+{
+    public CancellationToken CancellationToken { get; }
+
+    public void AddFile(string hintName, Action<System.IO.Stream> callback);
+}
+```
+
+This can be used by a generator author like so:
+
+```csharp
+internal sealed class Generator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var compilationSource = context.CompilationProvider;
+
+        context.RegisterArtifactOutput(compilationSource, (nspc, item) =>
+        {
+            nspc.AddFile("interop.js", (stream) =>
+            {
+                using var writer = new StreamWriter(stream);
+                writer.WriteLine("//output");
```

Comment:
The problem with having `CancellationToken` be on the `ArtifactProductionContext` is that accessing it here, which seems natural as you'd pass it to the `Stream` APIs, requires a closure and an allocation.